### PR TITLE
Add N-ary broadcasting operations.

### PR DESCRIPTION
**Pull request by hameerabbasi**
This PR adds N-ary broadcasting operations (in preparation for `where`) and simplifies code for the N-ary case.

Discussed in https://github.com/pydata/sparse/issues/1

### CODE REVIEW

**Comment by hameerabbasi** [comment]
cc @shoyer Your input would be valuable, if you have time.
cc @mrocklin Your input is welcome, as always.

File: sparse/coo.py [hunk 1]
```diff
@@ -2149,101 +2117,153 @@ def _elemwise_binary(func, self, other, *args, **kwargs):
...
-    assert isinstance(other, COO)
+def _match_coo(*args):
+    """
+    Matches the coordinates for any number of input :obj:`COO` arrays.
+    Equivalent to "sparse" broadcasting for all arrays.
 
-    result_shape = _get_broadcast_shape(self.shape, other.shape)
+    Parameters
+    ----------
+    args : tuple[COO]
```
Comment (mrocklin) [inline sparse/coo.py]: If you're trying for parametrized python type annotations then I think it's supposed to be standard to use capitalized types like `List[COO]` or `Tuple[np.ndarray]`

Comment (mrocklin) [reply sparse/coo.py]: I don't know though, this is somewhat new to me.

File: sparse/coo.py [hunk 2]
```diff
@@ -2034,113 +1988,127 @@ def _grouped_reduce(x, groups, method, **kwargs):
...
-        _get_reduced_coords(other_coords, other_shape,
-                            other_reduce_params)
-    other_reduced_linear = _linear_loc(other_reduced_coords, other_reduced_shape)
-    i = np.argsort(other_reduced_linear)
-    other_reduced_linear = other_reduced_linear[i]
-    other_coords = other_coords[:, i]
-    other_data = other_data[i]
+    # Filter out scalars as they are 'baked' into the function.
+    func = _posarg_partial(func, pos, posargs)
+    args = list(filter(lambda arg: not isscalar(arg), args))
```
Comment (mrocklin) [inline sparse/coo.py]: You might consider `toolz.remove` here

File: sparse/coo.py [hunk 3]
```diff
@@ -2034,113 +1988,127 @@ def _grouped_reduce(x, groups, method, **kwargs):
...
+    See Also
+    --------
+    :obj:`numpy.ufunc` : A similar Numpy construct. Note that any :code:`ufunc` can be used
+        as the :code:`func` input to this function.
+    """
+    # Because we need to mutate args.
+    args = list(args)
+    posargs = []
+    pos = []
+    for i in range(len(args)):
```
Comment (mrocklin) [inline sparse/coo.py]: You might consider `for i, arg in enumerate(args)`, which might be a bit more idiomatic for Python readers

File: sparse/coo.py [hunk 4]
```diff
@@ -1954,6 +1894,20 @@ def tril(x, k=0):
     return COO(coords, data, x.shape, x.has_duplicates, x.sorted)
 
 
+def _nary_match(*arrays):
```
Comment (mrocklin) [inline sparse/coo.py]: I'm not able to quickly figure out what this function does.  Can I ask you for a small docstring?  If possible I find small example sections in docstrings to be very helpful when learning codebases that others have written.

File: sparse/coo.py [hunk 5]
```diff
@@ -2034,113 +1988,127 @@ def _grouped_reduce(x, groups, method, **kwargs):
...
+    for mask in product([True, False], repeat=len(args)):
+        if not any(mask):
+            continue
 
-        coords_list.extend(other_unmatched_coords)
-        data_list.extend(other_unmatched_func)
+        ci, di = _unmatch_coo(func, args, mask, **kwargs)
+
+        coords_list.extend(ci)
+        data_list.extend(di)
```
Comment (mrocklin) [inline sparse/coo.py]: This confuses me and seems concerning.  I see that this was a main point of your conversation with @shoyer earlier.  I probably have some thinking to do on this problem before I'm able to reasonably comment on this.

**Comment by mrocklin** [comment]
I think that it would be good to see a more comprehensive test suite that fully explains the complexity of what we're trying to accomplish here.  I think that that will make it more clear as we discuss different possibilities here.  We might ask "why are we doing X" and the answer can be "see test_X".  I get the sense that you've thought deeply about this problem and know all of the problems that might arise.  It would be very valuable to encode that deep thinking and all of those corner cases into a test suite.

File: sparse/coo.py [hunk 6]
```diff
@@ -2511,8 +2512,39 @@ def _linear_loc(coords, shape, signed=False):
...
+
+        def __str__(self):
+            return str(func)
+
+        def __repr__(self):
+            return repr(func)
+
+        __doc__ = func.__doc__
+
+    return Partial()
```
[the remark this answers is not in the document]
Comment (hameerabbasi) [reply sparse/coo.py]: I could have, but our situation is slightly unique:

1. We're using `str(func)` for exceptions. `functools.wraps` doesn't work on that for all callables (e.g. `ufunc`s), and breaks a few docstrings. This leads to illegible names in exceptions like `_posarg_partial.<locals>.wrapper` (and the same for debugging).
2. We're replacing a number of arguments in different positions.

I guess I could turn it into a class rather than a decorator style function.

File: sparse/coo.py (hunk 1 above)
Comment (hameerabbasi) [reply sparse/coo.py]: `tuple` and `list` tend to work better with intersphinx and code type annotations, so I tend to prefer those. Of course, I could import in something, but then that gives me PEP8 failures as I don't use it in code, just in docstrings.

File: sparse/coo.py (hunk 2 above)
Comment (hameerabbasi) [reply sparse/coo.py]: I'd prefer not to introduce a dependency for something as simple as this.

**Comment by hameerabbasi** [comment]
I plan to make more comprehensive tests, yes. But the issue is some of the complexity can't be directly tested: For example, the optimizations are just that: Optimizations. We can design the tests so the optimizations are hit but we can't know that they kicked in without weird monkey-patching of some sort.

File: sparse/coo.py (hunk 4 above)
Comment (hameerabbasi) [reply sparse/coo.py]: Whoops, must have missed that one.

File: sparse/coo.py (hunk 3 above)
Comment (hameerabbasi) [reply sparse/coo.py]: Done!

File: sparse/coo.py [hunk 7]
```diff
@@ -2426,80 +2468,39 @@ def _elemwise_unary(func, self, *args, **kwargs):
...
+    coords : list[numpy.ndarray]
+        The input coordinates.
+    params : list[Union[bool, none]]
+        The broadcast parameters.
     Returns
     -------
-    matching_coords : np.ndarray
-        The coordinates of the output array for which both inputs will be nonzero.
+    numpy.ndarray
+        The broacasted coordinates.
```
[the remark this answers is not in the document]
Comment (hameerabbasi) [reply sparse/coo.py]: Done!

File: sparse/coo.py [hunk 8]
```diff
@@ -2426,80 +2468,39 @@ def _elemwise_unary(func, self, *args, **kwargs):
                sorted=self.sorted)
 
 
-def _get_matching_coords(coords1, coords2, shape1, shape2):
+def _get_nary_matching_coords(coords, params, shape):
```
[the remark this answers is not in the document]
Comment (hameerabbasi) [reply sparse/coo.py]: Done!

File: sparse/coo.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hameerabbasi) [reply sparse/coo.py]: I turned it into a callable class.

**Comment by hameerabbasi** [comment]
It seems there's a slight bug for number of inputs >2 and broadcasting, nothing unfixable, but will have to think a bit. I'm on it.

**Comment by mrocklin** [comment]
> I plan to make more comprehensive tests, yes. But the issue is some of the complexity can't be directly tested: For example, the optimizations are just that: Optimizations. We can design the tests so the optimizations are hit but we can't know that they kicked in without weird monkey-patching of some sort.

I think that there are probably a lot of correctness tests that could be written as well.  In #1 you discussed many situations that might arise for which a system like this would be necessary to catch.  Ideally we would encode all of those situations as tests to ensure that future developers don't change code to alter correct behavior here.

File: sparse/coo.py [hunk 9]
```diff
@@ -2034,113 +2006,124 @@ def _grouped_reduce(x, groups, method, **kwargs):
...
-        _get_reduced_coords(other_coords, other_shape,
-                            other_reduce_params)
-    other_reduced_linear = _linear_loc(other_reduced_coords, other_reduced_shape)
-    i = np.argsort(other_reduced_linear)
-    other_reduced_linear = other_reduced_linear[i]
-    other_coords = other_coords[:, i]
-    other_data = other_data[i]
+    # Filter out scalars as they are 'baked' into the function.
+    func = PositinalArgumentPartial(func, pos, posargs)
+    args = list(filter(lambda arg: not isscalar(arg), args))
```
Comment (shoyer) [inline sparse/coo.py]: optional: consider using a list comprehension instead

File: sparse/coo.py [hunk 10]
```diff
@@ -2149,101 +2132,153 @@ def _elemwise_binary(func, self, other, *args, **kwargs):
...
-    shape : tuple[int]
-        The shape corresponding to these coordinates.
-    result_shape : tuple[int]
-        The result broadcasting shape.
-    matched_idx : np.ndarray
-        The indices into the coords array where it matches with the other array.
-    matched_coords : np.ndarray
-        The overall coordinates that match from both arrays.
+    args : tuple[COO]
+        The input :obj:`COO` arrays.
```
Comment (shoyer) [inline sparse/coo.py]: add in `func`, `mask` and `**kwargs` to the docstring?

File: sparse/coo.py [hunk 11]
```diff
@@ -2149,101 +2132,153 @@ def _elemwise_binary(func, self, other, *args, **kwargs):
...
+    matched_args = [a for a, m in zip(args, mask) if m]
+    unmatched_args = [a for a, m in zip(args, mask) if not m]
 
-    unmatched &= nonzero
-    matched &= nonzero
+    matched_arrays = _match_coo(*matched_args)[1]
 
-    coords_list = []
-    data_list = []
+    pos, = np.where([not m for m in mask])
```
Comment (shoyer) [inline sparse/coo.py]: maybe use `np.flatnonzero()`?

**Comment by shoyer** [comment]
I agree with @mrocklin that a more extensive test suite is vital here. This logic is complicated and fixing bugs later will be hard. I haven't seriously tried to follow it yet.

I would suggest parametric tests verifying proper broadcasting with 2 or 3 arguments with:
- `*` for sparse broadcasting
- `+` for dense broadcasting
- an order dependent function (e.g., `-`) to verify you get order right
- with pathological values `np.infty` and `np.nan`

File: sparse/coo.py (hunk 9 above)
Comment (hameerabbasi) [reply sparse/coo.py]: Done!

File: sparse/coo.py (hunk 10 above)
Comment (hameerabbasi) [reply sparse/coo.py]: Done!

File: sparse/coo.py (hunk 11 above)
Comment (hameerabbasi) [reply sparse/coo.py]: This isn't really a numerical operation. I've converted it to a `tuple(generator comprehension)` form and avoided `np.where` altogether. The exact code is
```
pos = tuple(i for i, m in enumerate(mask) if not m)
```

**Comment by hameerabbasi** [comment]
I think this PR is now ready for a comprehensive review + merge.

File: sparse/coo.py (hunk 4 above)
Comment (mrocklin) [reply sparse/coo.py]: It looks like this function is no longer used.  Delete?

File: sparse/coo.py [hunk 12]
```diff
@@ -2034,113 +2006,125 @@ def _grouped_reduce(x, groups, method, **kwargs):
...
+        elif isscalar(arg) or (isinstance(arg, np.ndarray)
+                               and not arg.shape):
+            # Faster and more reliable to pass ()-shaped ndarrays as scalars.
+            if isinstance(arg, np.ndarray):
+                args[i] = arg[()]
+
+            pos.append(i)
+            posargs.append(args[i])
+        elif isinstance(arg, SparseArray) and not isinstance(arg, COO):
+            args[i] = COO(arg)
```
Comment (mrocklin) [inline sparse/coo.py]: This line doesn't get hit by tests.  Should we add a small DOK test?

File: sparse/coo.py [hunk 13]
```diff
@@ -2034,113 +2006,125 @@ def _grouped_reduce(x, groups, method, **kwargs):
...
+        elif not isinstance(arg, COO):
             raise ValueError("Performing this operation would produce "
                              "a dense result: %s" % str(func))
```
Comment (mrocklin) [inline sparse/coo.py]: Same here.  No test triggers this error-handling code.

File: sparse/coo.py [hunk 14]
```diff
@@ -2034,113 +2006,125 @@ def _grouped_reduce(x, groups, method, **kwargs):
...
-
-    # Start with an empty list. This may reduce computation in many cases.
-    data_list = []
-    coords_list = []
+    # Filter out scalars as they are 'baked' into the function.
+    func = PositinalArgumentPartial(func, pos, posargs)
+    args = [arg for arg in args if not isscalar(arg)]
+
+    if len(args) == 0:
+        return func(**kwargs)
```
Comment (mrocklin) [inline sparse/coo.py]: Also here.  No test operates on no args

File: sparse/coo.py [hunk 15]
```diff
@@ -2149,101 +2133,173 @@ def _elemwise_binary(func, self, other, *args, **kwargs):
...
+        ``return_midx`` is ``False``.
+    """
+    return_midx = kwargs.pop('return_midx', False)
+    cache = kwargs.pop('cache', None)
+
+    if kwargs:
+        raise ValueError('Unknown kwargs %s' % kwargs.keys())
+
+    if return_midx and (len(args) != 2 or cache is not None):
+        raise NotImplementedError('Matching only supported for two args, and no cache.')
```
Comment (mrocklin) [inline sparse/coo.py]: Do we still need this option?

File: sparse/tests/test_coo.py [hunk 16]
```diff
@@ -234,6 +234,113 @@ def test_elemwise_binary(func, shape):
...
+     ], lambda x, y, z: x - y + z),
+])
+def test_nary_broadcasting(shapes, func):
+    args = [sparse.random(s, density=0.5) for s in shapes]
+    dense_args = [arg.todense() for arg in args]
+
+    fs = sparse.elemwise(func, *args)
+    assert isinstance(fs, COO)
+
+    assert_eq(fs, func(*dense_args))
```
Comment (mrocklin) [inline sparse/tests/test_coo.py]: It would be nice to test and verify that we are not creating unnecessary zeroes in the data attribute.  We might either test that explicitly here, or we might put it into `assert_eq`.  I've gone ahead and pushed a commit to your branch that adds a check into `assert_eq`.  Please remove if you prefer not to add this here.

File: sparse/tests/test_coo.py [hunk 17]
```diff
@@ -234,6 +234,113 @@ def test_elemwise_binary(func, shape):
...
+    np.nan,
+    np.inf,
+    -np.inf
+])
+@pytest.mark.filterwarnings('ignore:invalid value')
+def test_nary_broadcasting_pathological(shapes, func, value):
+    def value_array(n):
+        ar = np.empty((n,), dtype=np.float_)
+        ar[:] = value
+        return ar
```
Comment (mrocklin) [inline sparse/tests/test_coo.py]: We might want just a few of the values to be pathological instead of all of them.

File: sparse/coo.py (hunk 15 above)
Comment (hameerabbasi) [reply sparse/coo.py]: No, we don't. I'm not omniscient, so I went ahead and added this check in case someone tried to trigger caching on `return_midx` (which we don't cache, it's never repeated); or tried to match indices for `len(args) != 2` (I'm not sure if we'll need this in the future, but we might, and it's useful to err rather than have it return incorrect results).

File: sparse/tests/test_coo.py (hunk 16 above)
Comment (hameerabbasi) [reply sparse/tests/test_coo.py]: I'd like to verify we don't create additional zeros for all our operations, so that seems like a rather useful addition.

Comment (hameerabbasi) [reply sparse/tests/test_coo.py]: Although I would prefer to use `np.count_nonzero`.

File: sparse/tests/test_coo.py (hunk 17 above)
Comment (hameerabbasi) [reply sparse/tests/test_coo.py]: I'll modify the test to match that.

File: sparse/coo.py (hunk 12 above)
Comment (hameerabbasi) [reply sparse/coo.py]: Added!

File: sparse/coo.py (hunk 13 above)
Comment (hameerabbasi) [reply sparse/coo.py]: Added a small test that hits this.

File: sparse/coo.py (hunk 14 above)
Comment (hameerabbasi) [reply sparse/coo.py]: Added another small test for this.

File: sparse/tests/test_coo.py (hunk 17 above)
Comment (hameerabbasi) [reply sparse/tests/test_coo.py]: Modified.

**Comment by hameerabbasi** [comment]
I've incorporated more or less all of your suggestions about coverage, with one exception (see comments!)

**Comment by mrocklin** [comment]
@shoyer do you have a chance to look at this?  "Nope" is a fine answer.

**Comment by hameerabbasi** [comment]
I'm guessing @shoyer doesn't work weekends. :-) If there's no reply or a "Nope" by the end of Monday, we can decide what to do next.

**Comment by shoyer** [comment]
It's a mixed bag on weekend, but this weekend my wife is away so I have time for open source :).

I'll take a look.

File: sparse/tests/test_coo.py [hunk 18]
```diff
@@ -250,7 +250,128 @@ def test_elemwise_trinary(func, shape):
...
+
+    assert_eq(fs, func(x, y, z))
+
+
+@pytest.mark.parametrize('shapes, func', [
+    ([
+         (2,),
+         (3, 2),
+         (4, 3, 2),
+     ], lambda x, y, z: (x + y) * z),
```
Comment (shoyer) [inline sparse/tests/test_coo.py]: Consider doing a full cross-product of shapes and functions here.

File: sparse/tests/test_coo.py [hunk 19]
```diff
@@ -250,7 +250,128 @@ def test_elemwise_trinary(func, shape):
...
+         (4,),
+         (4, 4),
+         (4, 4, 4),
+     ], lambda x, y, z: x + y - z),
+    ([
+         (4,),
+         (4, 4),
+         (4, 4, 4),
+     ], lambda x, y, z: x - y + z),
+])
```
Comment (shoyer) [inline sparse/tests/test_coo.py]: It would be good to add checks for a few more variations on the broadcasting logic to exercise the matching logic:
- Dimensions of size 1, e.g., `(3, 1)` + `(3, 4)` -> `(3, 4)`
- Output shapes that don't match one of the inputs, e.g., `(3, 1)` + `(1, 4)` -> `(3, 4)`.
- Outputs that require matching across three inputs, e.g., `(1, 1, 2)` + `(1, 3, 1)` + `(4, 1, 1)` -> `(4, 3, 2)`.

File: sparse/tests/test_coo.py (hunk 16 above)
Comment (shoyer) [reply sparse/tests/test_coo.py]: On a related note, maybe it would be useful to add a test that verifies "sparse" broadcasting is actually done in a sparse way?

I think you could do this by mocking the underlying functions (e.g., `np.mul`) and then verifying that the calls match expectations.

File: sparse/tests/test_coo.py (hunk 19 above)
Comment (hameerabbasi) [reply sparse/tests/test_coo.py]: The first two were already covered in `test_broadcasting`. I renamed that to `test_binary_broadcasting` and moved it closer to these.

The third, I also added.

File: sparse/tests/test_coo.py (hunk 18 above)
Comment (hameerabbasi) [reply sparse/tests/test_coo.py]: Done!

**Comment by hameerabbasi** [comment]
I can't seem to be able to respond to your "sparse" broadcasting comment, so I'm responding here.

I monkey-patched one of our own functions and verified the behavior is correct there. I also verified `operator.add` takes the 'dense' path.

**Comment by mrocklin** [comment]
> I can't seem to be able to respond to your "sparse" broadcasting comment, so I'm responding here.

Yeah, I was trying to do that as well.  I haven't seen that before

Is checking for the right number of non-zeros in the output not sufficient?  Do we have code paths that would re-sparsify a dense intermediate result?

**Comment by hameerabbasi** [comment]
We're actually talking about the "optimized" code path for things like `operator.mul` where it only calculates full matches and not partial matches.

**Comment by mrocklin** [comment]
Ah, happy to retract my comment

**Comment by hameerabbasi** [comment]
No, let's leave it up for other people who can't follow the terminology.

**Comment by hameerabbasi** [comment]
Are there any further comments or is this good to merge?

**Comment by mrocklin** [comment]
Merged!
